### PR TITLE
[BEAM-7945] Allow runner to configure semi_persist_dir which is used …

### PR DESCRIPTION
**Pull request by sunjincheng121**
Currently "semi_persist_dir" is not configurable. This may become a problem in certain scenarios. For example, the default value of "semi_persist_dir" is "/tmp" (https://github.com/apache/beam/blob/master/sdks/python/container/boot.go#L48) in Python SDK harness. When the environment type is "PROCESS", the disk of "/tmp" may be filled up and unexpected issues will occur in production environment. 

So, This pull request makes the semi_persist_dir configurable through adding a new PipelineOption(RemoteEnvironmentOptions).The Pipeline option will be passed to the `DefaultJobBundleFactory` and then be used in each EnvironmentFactory(docker, process, external and embedded).

For details of the discussion can be found in [1].

[1] https://lists.apache.org/list.html?dev@beam.apache.org:lte=1M:%5BDISCUSS%5D%20Turn%20%60WindowedValue

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by sunjincheng121** [comment]
I appreciate if you have time to look up the changes @robertwb @mxm :)

File: model/fn-execution/src/main/proto/beam_fn_api.proto [hunk 1]
```diff
@@ -815,6 +815,7 @@ message StartWorkerRequest {
   org.apache.beam.model.pipeline.v1.ApiServiceDescriptor logging_endpoint = 3;
   org.apache.beam.model.pipeline.v1.ApiServiceDescriptor artifact_endpoint = 4;
   org.apache.beam.model.pipeline.v1.ApiServiceDescriptor provision_endpoint = 5;
+  string semi_persist_dir = 6;
```
Comment (mxm) [inline model/fn-execution/src/main/proto/beam_fn_api.proto]: Should this be dynamic or rather configured up front for the worker pool?

Comment (sunjincheng121) [reply model/fn-execution/src/main/proto/beam_fn_api.proto]: Great to have your suggestions. :)
Maybe I have not understood your idea. If we configure the dir for the whole pool, we 
may loose the flexibility that different jobs may set different semi_persist_dir for the workers in the same worker pool?

Comment (mxm) [reply model/fn-execution/src/main/proto/beam_fn_api.proto]: I'm not sure whether this flexibility is desired. I could imagine that the person who starts the worker pool does not want arbitrary persist directories, but rather rather a fixed one.

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/DefaultJobBundleFactory.java [hunk 2]
```diff
@@ -92,18 +92,18 @@
   private final int environmentExpirationMillis;
 
   public static DefaultJobBundleFactory create(JobInfo jobInfo) {
+    PipelineOptions pipelineOption =
```
Comment (tweise) [inline runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/DefaultJobBundleFactory.java]: "pipelineOptions"

File: model/fn-execution/src/main/proto/beam_fn_api.proto (hunk 1 above)
Comment (tweise) [reply model/fn-execution/src/main/proto/beam_fn_api.proto]: This should not be added here. Pipeline options are provided through the provisioning endpoint.

Comment (sunjincheng121) [reply model/fn-execution/src/main/proto/beam_fn_api.proto]: Oh, Yes, I see, this is useless changes. `pipeline_options` already defined in the `ProvisionInfo`.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/options/RemoteEnvironmentOptionsTest.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+/** Tests for {@link RemoteEnvironmentOptions}. */
+@RunWith(JUnit4.class)
+public class RemoteEnvironmentOptionsTest {
+
+  @Test
+  public void testSemiDirectory() {
+    RemoteEnvironmentOptions options = PipelineOptionsFactory.as(RemoteEnvironmentOptions.class);
+    String semiDir = "/ab/cd";
+    options.setSemiPersistDir(semiDir);
+    assertEquals(semiDir, options.getSemiPersistDir());
```
Comment (mxm) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/options/RemoteEnvironmentOptionsTest.java]: This should test the default.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/RemoteEnvironmentOptions.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
+
+/** Options that are used to control configuration of the remote environment. */
+@Experimental
+@Hidden
+public interface RemoteEnvironmentOptions extends PipelineOptions {
+
+  @Description("Local semi-persistent directory")
+  @Default.String("/tmp")
```
Comment (mxm) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/RemoteEnvironmentOptions.java]: Should this be `System.getProperty('java.io.tmpdir')`?

Comment (sunjincheng121) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/RemoteEnvironmentOptions.java]: Currently, we keep the same as the default value of other default configuration, such as:`boot.go`.
- https://github.com/apache/beam/blob/d21bbaf4c70986c2dbdbe8f6fce35b2b2cb4843d/sdks/go/container/boot.go#L41
- https://github.com/apache/beam/blob/d21bbaf4c70986c2dbdbe8f6fce35b2b2cb4843d/sdks/python/container/boot.go#L51

So, how about we keep using `/tmp` as default value ?

Comment (mxm) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/RemoteEnvironmentOptions.java]: Let's keep the existing default.

**Comment by sunjincheng121** [comment]
Thanks for the review @mxm !
I have update the PR according your comments. I appreciate if you can have a another look.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/RemoteEnvironmentOptions.java (hunk 4 above)
Comment (tweise) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/RemoteEnvironmentOptions.java]: I think the default should be null, so that the environment can pick its suitable tmp directory when nothing is specified by the user.

**Comment by sunjincheng121** [comment]
@mxm @tweise  Thanks a lot for the review. Sorry that I missed the comments from @tweise.
I have updated the PR, would be great if you can take another look.

**Comment by sunjincheng121** [comment]
Run Java PreCommit

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerEnvironmentFactory.java [hunk 5]
```diff
@@ -132,20 +133,30 @@ public RemoteEnvironment createEnvironment(Environment environment) throws Excep
...
+
+    String semiPersistDir = pipelineOptions.as(RemoteEnvironmentOptions.class).getSemiPersistDir();
+    ImmutableList.Builder<String> argsBuilder =
+        ImmutableList.<String>builder()
+            .add(String.format("--id=%s", workerId))
+            .add(String.format("--logging_endpoint=%s", loggingEndpoint))
+            .add(String.format("--artifact_endpoint=%s", artifactEndpoint))
+            .add(String.format("--provision_endpoint=%s", provisionEndpoint))
+            .add(String.format("--control_endpoint=%s", controlEndpoint));
+    if (semiPersistDir != null) {
```
Comment (tweise) [inline runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerEnvironmentFactory.java]: Isn't this provided to the environment through the provision endpoint?

Comment (tweise) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerEnvironmentFactory.java]: So we essentially pass the same piece of information to the worker twice: As entry point argument and then again within the pipeline options. It needs to be done this way due to the container contract. Would be nice to revisit in the future.

Comment (mxm) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerEnvironmentFactory.java]: That's right. This is redundant here.

Comment (mxm) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerEnvironmentFactory.java]: Actually, the semi_persist_dir is not inferred from the pipeline options in the bootloader code. Like you said, it has to be this way currently, but it would be nice to not duplicate this information in the future.

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerEnvironmentFactory.java [hunk 6]
```diff
@@ -132,20 +133,30 @@ public RemoteEnvironment createEnvironment(Environment environment) throws Excep
...
-        ImmutableList.of(
-            String.format("--id=%s", workerId),
-            String.format("--logging_endpoint=%s", loggingEndpoint),
-            String.format("--artifact_endpoint=%s", artifactEndpoint),
-            String.format("--provision_endpoint=%s", provisionEndpoint),
-            String.format("--control_endpoint=%s", controlEndpoint));
+    Boolean retainDockerContainer =
+        pipelineOptions.as(ManualDockerEnvironmentOptions.class).getRetainDockerContainers();
+    if (!retainDockerContainer) {
+      dockerOptsBuilder.add("--rm");
```
Comment (ibzib) [inline runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerEnvironmentFactory.java:139]: Why was this added in this PR? It seems orthogonal to `semi_persist_dir`. I believe this is a regression; perhaps a rebasing error?

Comment (mxm) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerEnvironmentFactory.java:139]: Indeed looks like a rebasing error. A bit tricky one to get right because we changed the way the container removal worked before the rebasing was done. We were using the `--rm` initially but before the rebase we changed it to remove the container explicitly via `docker remove <container_id>`.

Comment (mxm) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerEnvironmentFactory.java:139]: Fix: https://github.com/apache/beam/pull/11303
